### PR TITLE
Only calculate tax in the 'confirm' state.

### PR DESCRIPTION
**Pull request by jordan-brough**
Tax calculation is expensive so we're going to stop calculating tax prior to the
'confirm' step.

In the places where we previously calculated SalesOrders (pre-confirm,
non-persisted tax estimates) instead clear out any prevoius tax data. This is to
handle the case where an order moves backward from the 'confirm' step. e.g. if
the user adds something else to their cart.

### CODE REVIEW

File: app/models/spree/address_decorator.rb [hunk 1]
```diff
@@ -3,7 +3,7 @@
 
   def avatax_compute_tax
```
Comment (athal7) [inline app/models/spree/address_decorator.rb]: does it make sense to rename this now?

File: app/models/spree/order_contents_decorator.rb [hunk 2]
```diff
@@ -1,19 +1,19 @@
...
   end
 
   def remove_with_avatax(*args)
     remove_without_avatax(*args).tap do
-      SpreeAvatax::SalesOrder.generate(order)
+      SpreeAvatax::SalesShared.reset_tax_attributes(order)
     end
   end
 
   def update_cart_with_avatax(params)
```
Comment (athal7) [inline app/models/spree/order_contents_decorator.rb]: i think this one gets called post-complete when an address is updated, which would then delete tax adjustments

File: app/models/spree_avatax/sales_shared.rb [hunk 3]
```diff
@@ -95,6 +95,51 @@ def avatax_id(record)
       "#{record.class.name}-#{record.id}"
     end
 
+
+    # Clears previously-set tax attributes from an order, if any, unless the
+    # order has already been completed.
+    #
+    # @param order [Spree::Order] the order
+    def reset_tax_attributes(order)
+      return if order.completed?
```
Comment (athal7) [inline app/models/spree_avatax/sales_shared.rb]: ah i see your check here, disregard my other comment about the order being completed

File: app/models/spree_avatax/sales_shared.rb [hunk 4]
```diff
@@ -95,6 +95,51 @@ def avatax_id(record)
...
+
+      destroyed_adjustments = bench('reset_tax', 'adjustments') do
+        order.all_adjustments.tax.destroy_all
+      end
+
+      return if destroyed_adjustments.empty?
```
Comment (athal7) [inline app/models/spree_avatax/sales_shared.rb]: nice! this should have some performance benefits

File: app/models/spree/address_decorator.rb (hunk 1 above)
Comment (jordan-brough) [inline app/models/spree/address_decorator.rb]: yup, good call.

Comment (jordan-brough) [inline app/models/spree/address_decorator.rb]: updated.

**Comment by athal7** [comment]
:+1:

**Comment by jordan-brough** [comment]
will merge after we finish testing it out

**Pull request closed without merging**